elastic.go

```diff
@@ -2,38 +2,76 @@ package main
 
 import (
 	"bytes"
-	"compress/zlib"
 	"context"
+	"crypto/rand"
 	"crypto/tls"
 	"crypto/x509"
+	"encoding/base64"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
+	"path/filepath"
 	"reflect"
+	"strconv"
 	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
 
-	elastic "github.com/elastic/go-elasticsearch/v7"
+	"github.com/cenkalti/backoff"
+	"github.com/elastic/go-elasticsearch/v7"
 	"github.com/elastic/go-elasticsearch/v7/esutil"
-	log "github.com/sirupsen/logrus"
 	"github.com/tidwall/gjson"
+
+	log "github.com/sirupsen/logrus"
 )
 
 // ElasticConfig is a Struct that holds ElasticSearch config
-type ElasticConfig struct {
+type elasticConfig struct {
+	host       string
+	port       int
 	user       string
 	password   string
-	verifyPeer bool
+	pkiAuth    bool
```

Comment:
This name is vague to me. PKI auth sounds like you can authenticate a user to elasticsearch using a certificate. This is a premium feature of ES which we don't have. Perhaps a better name could be `useTLS`?

Reply:
Nope.
For HTTPS you only need a CA to validate against, that can be system global or a specific CA for just ES.
If you connect to ES using a client certificate you will be using it for authentication.

The choice between HTTP and HTTPS depends on the URI you are calling.

Reply:
For client authentication, but not authorization. So again this is not totally clear.

Reply:
It serves the same purpose as username:password.

Reply:
I guess you are referring to this:
https://www.elastic.co/guide/en/elasticsearch/reference/current/pki-realm.html

Reply:
Removed